IDGEN-67 Follow Up

cnagadya opened the pull request:
## JIRA TICKET NAME:

[IDGEN-94 Follow Up](https://issues.openmrs.org/browse/IDGEN-94)

## SUMMARY:
Fixes the following bugs / issues in the LogEntry classes that were created in [IDGEN-67](https://github.com/openmrs/openmrs-module-idgen/pull/44):

- Removes the source property getter
- Bug in the search fields **source** and **generatedBy**
- Move the **resource** and **searchhandler** tests from web folder to rest folder

dkayiwa commented on omod/src/test/java/org/openmrs/module/idgen/web/search/LogEntrySearchHandlerTest.java:
```diff
@@ -82,7 +82,46 @@ public void getSearchConfig_shouldReturnLogEntryByDateGenerated() throws Excepti
+	}
+
+	@Test
+	public void getSearchConfig_shouldReturnLogEntryByGeneratedByandComment() throws Exception {
```
Did you intentionally have some of your "And" s start with capital A while others start with small a?

cnagadya replied:
Changed them all to start with "A". Also pushed the failing tests

dkayiwa commented on omod/src/test/java/org/openmrs/module/idgen/web/search/LogEntrySearchHandlerTest.java:
```diff
@@ -82,7 +82,65 @@ public void getSearchConfig_shouldReturnLogEntryByDateGenerated() throws Excepti
+		Assert.assertNotNull(result);
+		Assert.assertEquals(2, Util.getResultsSize(result));
+	}
+		@Test
```
Did you intentionally format this like so?

cnagadya replied:
That was an oversight on my side. Going to fix it now

dkayiwa commented:
@cnagadya have you looked at each of these failing tests?

cnagadya commented:
Yes, was trying to replicate [the issue ](https://talk.openmrs.org/t/bug-in-the-advanced-search-for-log-entries/13912/4)that I had raised on talk

dkayiwa commented:
@cnagadya i ask again. Have you looked at each of the failing tests in this pull request?

cnagadya commented:
Didnt realise the data I added was causing [_shouldSearchAndReturnAListOfLogEntriesWhoseCommentsContain_](https://github.com/cnagadya/openmrs-module-idgen/blob/9f4edcc1c0f044ecc73e755793f1cfae422fc8fe/omod/src/test/java/org/openmrs/module/idgen/web/controller/LogEntryControllerTest.java) and [_getSearchConfig_shouldReturnLogEntryByComment_](https://github.com/cnagadya/openmrs-module-idgen/blob/d4fe987721fc2ae34434670591d335ea2b6e4799/omod/src/test/java/org/openmrs/module/idgen/web/search/LogEntrySearchHandlerTest.java#L57-L61) tests to fail. Fixed it now

dkayiwa commented:
@cnagadya so now how many tests do you have left failing? And have you looked at them to confirm they are properly written according to the code expectations?

cnagadya commented:
Two tests are failing now 
[getSearchConfig_shouldReturnNoLogEntriesForInvalidGeneratedBy](https://github.com/cnagadya/openmrs-module-idgen/blob/d4fe987721fc2ae34434670591d335ea2b6e4799/omod/src/test/java/org/openmrs/module/idgen/web/search/LogEntrySearchHandlerTest.java#L131-L137)
[getSearchConfig_shouldReturnNoLogEntriesForInvalidSource](https://github.com/cnagadya/openmrs-module-idgen/blob/d4fe987721fc2ae34434670591d335ea2b6e4799/omod/src/test/java/org/openmrs/module/idgen/web/search/LogEntrySearchHandlerTest.java#L140-L147) just like I had indicated in the [talk post](https://talk.openmrs.org/t/bug-in-the-advanced-search-for-log-entries/13912/4)

dkayiwa commented:
@cnagadya to give an example of what am driving at, as for the first test, why do you expect it to return zero results, yet it has matching comments?

cnagadya commented:
Because the generatedBy is wrong. As long as one of the search fields doesnt match what is in the records, it should return no results. At least that is what I thought would happen

dkayiwa commented:
@cnagadya can you show me the section of code that you wrote to support that?

cnagadya commented:
In the [searchhandler](https://github.com/cnagadya/openmrs-module-idgen/blob/121bb50be3ac27f3c476806f6c662ac8479f3680/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L50-L55), I get the values for each of the search fields filled in by the user. For the [source](https://github.com/cnagadya/openmrs-module-idgen/blob/121bb50be3ac27f3c476806f6c662ac8479f3680/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L57) and [generatedBy](https://github.com/cnagadya/openmrs-module-idgen/blob/121bb50be3ac27f3c476806f6c662ac8479f3680/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L60) the provided values are uuids so I get the corresponding objects if they are not null then I pass the values to the [getLogEntries](idgen/blob/121bb50be3ac27f3c476806f6c662ac8479f3680/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L65) that is defined in [IdentifierSourceService](https://github.com/openmrs/openmrs-module-idgen/blob/master/api/src/main/java/org/openmrs/module/idgen/service/IdentifierSourceService.java#L245-L246) class which returns a list of the corresponding log entries

dkayiwa commented:
@cnagadya what happens if some of them are null?

cnagadya commented:
My understanding is that particular parameter will be passed as null. **getLogEntries** takes exactly 6 parameters that can either be **null** or **type** _IdentifierSource, Date , Date, String , User_ and  _String_ for _source, fromDate,  toDate, identifier, generatedBy_ and _comment_ respectively

dkayiwa commented:
What happens if some are null while others are not null? Can that return some results?

cnagadya commented:
Yes

dkayiwa commented:
@cnagadya so do your failing tests assume some results can be returned, or do they assume none?

cnagadya commented:
The expectation is that as long as one of the searchfields is wrong, no log entries are returned. In this case, I am entering invalid values for generatedBy or source. 
When a wrong value is entered for identifier (or comments), the app works as expected as shown in [this test](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/test/java/org/openmrs/module/idgen/web/controller/LogEntryControllerTest.java#L156-L161) even though other parameters are right

dkayiwa commented:
Can you point me to that exact section of code which ensures that if for instance source is not correct, no results are returned?

cnagadya commented:
@dkayiwa that is exactly where I was going wrong! Updated the how the [source](https://github.com/cnagadya/openmrs-module-idgen/blob/IDGEN-94/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L61-#L63) and [generatedBy](https://github.com/cnagadya/openmrs-module-idgen/blob/IDGEN-94/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L65-L67) are handled in the search handler

dkayiwa commented on omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java:
```diff
@@ -58,13 +58,19 @@ public PageableResult search(RequestContext context) throws ResponseException {
+        else if (generatedBy != null && user == null) {
+            return new EmptySearchResult();
+        } 
+        
```
What is the use of these empty lines?

cnagadya replied:
On one of my former PRs , one of the developers recommended that I should leave an empty line between code blocks

dkayiwa replied:
Empty lines between if and elses???

dkayiwa commented:
@cnagadya why did you decide to do this for only source and generatedBy?

cnagadya commented:
For the _identifier_, _comment_, _toDate_ and _fromDate_ the values provided by the user are used as is or converted to date format and passed to the **getLogEntries** method. However, for _source_ and _generatedBy_ the values are expected to be uuids that are passed to the **getIdentifierSourceByUuid** or **getUserByUuid** methods to return the associated objects or null if none is found. So I am checking to see if a user entered a value but their is no object with the given uuid

dkayiwa commented:
Would you expect any results if some one provides an invalid date?

cnagadya commented:
If the entered value is not in a date format, a 400 response is sent

cnagadya commented:
I am using the [java Date Util](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L15) to convert the values entered for [toDate](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L59) and [fromDate](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/main/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandler.java#L58) ant it automates the date validation.

[Here](https://drive.google.com/file/d/0Bz5uHaU8KGBPVnRFQ3BubmI1Nmc/view?usp=sharing) is a sample response for an invalid date entered

dkayiwa commented:
@cnagadya do you have a test which confirms that if an invalid date is entered, no results will be returned, even if the rest of the parameters are correct?

cnagadya commented:
I have added a [test](https://github.com/cnagadya/openmrs-module-idgen/blob/4f310fde58cbcd2d3145008aac8de334577c133c/omod/src/test/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandlerTest.java#L150-L158) for when there are no log entries within the specified range.

dkayiwa commented on omod/src/test/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandlerTest.java:
```diff
@@ -82,7 +82,78 @@ public void getSearchConfig_shouldReturnLogEntryByDateGenerated() throws Excepti
 	public void getSearchConfig_shouldReturnLogEntryByGeneratedBy() throws Exception {
 		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
 		req.addParameter("generatedBy", USER_UUID);
+		req.addParameter("comment", "New");
```
The package of this class does not match its folder structure. Can you look into that?

dkayiwa replied:
The same goes for LogEntryResourceTest

cnagadya replied:
Changed the packages. Its because I moved the both tests from the web to rest folder and I forgot to update the packages accordingly

dkayiwa commented on omod/src/test/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandlerTest.java:
```diff
@@ -82,7 +82,78 @@ public void getSearchConfig_shouldReturnLogEntryByDateGenerated() throws Excepti
+	@Test
+	public void getSearchConfig_shouldReturnNoLogEntriesIfDateRangeIsWrong() throws Exception {
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
+		req.addParameter("fromDate", "2017-10-21T12:00:00.433");
```
I mean invalid dates. As in strings which cannot be parsed into dates.

cnagadya replied:
The above test if for date range ie if no log entries were generated after the specified **fromDate**. I have added a [test](https://github.com/cnagadya/openmrs-module-idgen/blob/aae65f4f52c23a0acfee80a0e0f524ffd3067502/omod/src/test/java/org/openmrs/module/idgen/rest/search/LogEntrySearchHandlerTest.java#L161-L169) for invalid date

dkayiwa replied:
Since you are throwing an exception, the method name should change to reflect that.

dkayiwa replied:
Since you need to test to ensure that the rest client gets an error message which gives them a clue about what the invalid field is, can you use this? https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/api/OrderServiceTest.java#L149-L150

cnagadya replied:
Thanks for sharing the resource. I have made the recommended adjustments to the test
